libcloud/httplib_ssl.py

```diff
@@ -141,213 +142,99 @@ def _parse_proxy_url(self, proxy_url):
-
-        # otherwise, create a connection and verify the hostname
-        # use socket.create_connection (in 2.6+) if possible
-        if getattr(socket, 'create_connection', None):
-            sock = socket.create_connection((self.host, self.port),
-                                            self.timeout)
-        else:
-            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            sock.connect((self.host, self.port))
-
-        # Activate the HTTP proxy
-        if self.http_proxy_used:
-            self._activate_http_proxy(sock=sock)
-
-        ssl_version = libcloud.security.SSL_VERSION
-
-        # If we support SNI, use SSLContext and tls_context.wrap_socket()
-        # else revert to older behaviour with ssl.wrap_socket()
-        # Note: This feature is only available in Python 2.7.9 and
-        # Python >= 3.2
-        has_sni = getattr(ssl, 'HAS_SNI', False)
-
-        if has_sni:
-            self.tls_context = ssl.SSLContext(ssl_version)
-            self.tls_context.verify_mode = ssl.CERT_REQUIRED
-
-            if self.cert_file and self.key_file:
-                self.tls_context.load_cert_chain(
-                    certfile=self.cert_file,
-                    keyfile=self.key_file,
-                    password=None)
-
-            if self.ca_cert:
-                self.tls_context.load_verify_locations(cafile=self.ca_cert)
-
-            try:
-                self.sock = self.tls_context.wrap_socket(
-                    sock,
-                    server_hostname=self.host,
-                )
-            except:
-                exc = sys.exc_info()[1]
-                exc = get_socket_error_exception(ssl_version=ssl_version,
-                                                 exc=exc)
-                raise exc
-        else:
-            # SNI support not available
-            try:
-                self.sock = ssl.wrap_socket(
-                    sock,
-                    self.key_file,
-                    self.cert_file,
-                    cert_reqs=ssl.CERT_REQUIRED,
-                    ca_certs=self.ca_cert,
-                    ssl_version=ssl_version
-                )
-            except:
-                exc = sys.exc_info()[1]
-                exc = get_socket_error_exception(ssl_version=ssl_version,
-                                                 exc=exc)
-                raise exc
-
-        cert = self.sock.getpeercert()
-
-        # Verify Hostname
-        try:
-            match_hostname(cert, self.host)
-        except CertificateError:
-            e = sys.exc_info()[1]
-            raise ssl.SSLError('Failed to verify hostname: %s' % (str(e)))
+        self.session.timeout = kwargs.get('timeout', 60)
+
+    def request(self, method, url, body=None, headers=None, raw=False,
+                stream=False):
+        self.response = self.session.request(
+            method=method.lower(),
+            url=''.join([self.host, url]),
```

Comment:
ha. that found an error in morph_action_hook when the Connection object is instantiated from a URL. It would have a double slash. Pretty sure that would apply for <1.5.0